### PR TITLE
Changes required to support `jax` version 0.4.33

### PR DESCRIPTION
**Pull request by bwohlberg**
Changes required to support `jax` version 0.4.33, switch to using the new `orbax-checkpoint`  `CheckpointManager` API, and various other improvements.

### CODE REVIEW

File: examples/scripts/deconv_odp_train_foam1.py [hunk 1]
```diff
@@ -66,6 +66,11 @@
 
 import jax
 
+try:
+    from jax.extend.backend import get_backend  # introduced in jax 0.4.33
+except ImportError:
+    from jax.lib.xla_bridge import get_backend
+
```
Comment (Michael-T-McCann) [inline examples/scripts/deconv_odp_train_foam1.py:73]: consider making this a function giving that it is repeated several times

File: scico/numpy/_blockarray.py [hunk 2]
```diff
@@ -174,10 +174,18 @@ def prop_ba(self):
 def _da_method_wrapper(method_name):
     method = getattr(Array, method_name)
 
-    if method.__name__ is None:
+    if method.__name__ is None or method.__qualname__ is None:
         return method
```
Comment (Michael-T-McCann) [inline scico/numpy/_blockarray.py:178]: this is no longer needed given the later logic?

**Review by Michael-T-McCann — CHANGES_REQUESTED** [review]

File: examples/scripts/deconv_odp_train_foam1.py (hunk 1 above)
Comment (bwohlberg) [reply examples/scripts/deconv_odp_train_foam1.py:73]: I considered that, but it's going to be removed eventually (once the minimum supported version moves past 0.4.31), so it seemed more appropriate to do this.

File: scico/numpy/_blockarray.py (hunk 2 above)
Comment (bwohlberg) [reply scico/numpy/_blockarray.py]: Good point - fixed.

**Review by Michael-T-McCann — APPROVED** [review]

**Pull request merged**